### PR TITLE
Front-end basics for making API calls

### PR DESCRIPTION
**Pull request by lemald**
Asana Task: arbitrarily started as part of [Badge tap action via NFC API](https://app.asana.com/0/1200273269966439/1207259534606015/f)

@mathcolo sorry if any of this overlaps with your work on the [other part of the operator sign-in flow that will also have to get data from the backend](https://app.asana.com/0/1200273269966439/1207373383816236/f). I didn't go too far with my work so that if there's duplication we can sync up and take care of resolving it without too much effort.

Also one quick note that this is currently based off of the `lem-nfc-basics` branch because of some test utilities. Once #35 is merged I'll rebase this onto `main`.

Checklist

- Tests:
  - `(X)` Has tests
  - `( )` Doesn't need tests
  - `( )` Tests deferred (with justification)
- Product/Design sign off:
  - `( )` Okayed the plan for the feature (e.g. the design files, or the Asana task)
  - `( )` Reviewed the feature as implemented (e.g. on dev-green, or saw screenshots)
  - `(X)` No review needed

### CODE REVIEW

File: js/browser.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export const fetch: typeof window.fetch = window.fetch;
```
Comment (mathcolo) [inline js/browser.ts:1]: Can TypeScript not narrow/infer (whatever we call it?) these?

File: js/api.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    fetch(url)
+      .then((response) => {
+        if (response.status === 200) {
+          return response.json();
+        } else if (response.status === 401) {
+          reload();
+
+          throw new Error("Unauthenticated");
+        } else {
+          // TODO: handle various errors
```
Comment (mathcolo) [inline js/api.ts:31]: In the meantime, could we throw an `Unhandled response code` error or something?

File: js/browser.ts (hunk 1 above)
Comment (lemald) [reply js/browser.ts]: Looks like it can, I forget why I originally included that. ecbdcac

File: js/api.ts (hunk 2 above)
Comment (lemald) [reply js/api.ts]: Good point. I've made some improvements in 6f8d9b8. Note that throwing inside a promise just causes it to be rejected and the `throw` doesn't bubble up (which, no pun intended, threw me off when I was writing tests).

File: js/api.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        const jsonObject = z.object({ data: z.unknown() }).parse(json);
+
+        const data = RawData.parse(jsonObject.data);
+
+        const parsedData = parser(data);
+
+        setResult({ status: "ok", result: parsedData });
+      })
+      .catch((e: unknown) => {
+        setResult({ status: "error", error: e });
```
Comment (mathcolo) [inline js/api.ts:44]: Could we `console.error` this just so we see it as well?

**Review by mathcolo — APPROVED** [review]
One comment above, but it's non-blocking.

File: js/api.ts (hunk 3 above)
Comment (lemald) [reply js/api.ts:44]: I actually think sending an error to Sentry would be more useful, so I added a note to [the ticket](https://app.asana.com/0/616151179860796/1207620537833336/f) for frontend Sentry support. Does that sounds good to you?

**Pull request merged**